### PR TITLE
rgw: Introducing SSE-S3 RGW options and using them

### PR DESCRIPTION
**Pull request by prasakri**
Added SSE-S3 options for the Vault as the backend.

Signed-off-by: Priya Sehgal <priya.sehgal@flipkart.com>

TODO:
- [ ] per-upload encryption with header `x-amz-server-side-encryption: AES256` (and s3test cases)
- [ ] configuration to choose between per-bucket and per-user encryption keys (default to per-bucket)
- [ ] in teuthology, run sse-s3 test cases under vault transit configuration

### CODE REVIEW

File: src/rgw/rgw_kms.cc [hunk 1]
```diff
@@ -159,16 +159,35 @@ add_name_val_to_obj(const char *n, std::string &v, rapidjson::GenericValue<E,A>
...
+  virtual std::string & backend() = 0;
+  virtual std::string & addr() = 0;
+  virtual std::string & auth() = 0;
+  virtual std::string & k_namespace() = 0;
+  virtual std::string & prefix() = 0;
+  virtual const std::string & secret_engine() = 0;
+  virtual std::string & ssl_cacert() = 0;
+  virtual std::string & ssl_clientcert() = 0;
+  virtual std::string & ssl_clientkey() = 0;
+  virtual std::string & token_file() = 0;
```
[the remark this answers is not in the document]
Comment (priyagpssehgal) [reply src/rgw/rgw_kms.cc]: Sure, somehow I missed these.

**Comment by priyagpssehgal** [comment]
> looking at our rgw/crypt suite for this sse stuff, i'm wondering how best to integrate the new sse-s3 tests from [ceph/s3-tests#407](https://github.com/ceph/s3-tests/pull/407)
> 
> ```
> ~/ceph $ ls qa/suites/rgw/crypt/2-kms
> barbican.yaml  kmip.yaml  testing.yaml  vault_kv.yaml  vault_old.yaml  vault_transit.yaml
> ```
> 
> we test all of the sse-kms backends here, but sse-s3 tests will only work with the vault-transit backend. i see two options here:
> 
> 1. add this new `rgw_crypt_sse_s3_backend` configuration stuff directly to `vault_transit.yaml`, and only run the new sse-s3 tests in this configuration (using some extra s3tests filter in addition to `@attr('encryption')`)
> 2. outside of `2-kms/`, always deploy a vault instance and configure `rgw_crypt_sse_s3_backend` accordingly. this way we can always run the sse-s3 tests, along with the sse-kms tests against whatever backend is chosen from `2-kms`. i'm just not sure whether we can configure a single vault instance with transit enabled for sse-s3, but with vault_kv.yaml for sse-kms
> 
> what do you think @mdw-at-linuxbox?



> looking at our rgw/crypt suite for this sse stuff, i'm wondering how best to integrate the new sse-s3 tests from [ceph/s3-tests#407](https://github.com/ceph/s3-tests/pull/407)
> 
> ```
> ~/ceph $ ls qa/suites/rgw/crypt/2-kms
> barbican.yaml  kmip.yaml  testing.yaml  vault_kv.yaml  vault_old.yaml  vault_transit.yaml
> ```
> 
> we test all of the sse-kms backends here, but sse-s3 tests will only work with the vault-transit backend. i see two options here:
> 
> 1. add this new `rgw_crypt_sse_s3_backend` configuration stuff directly to `vault_transit.yaml`, and only run the new sse-s3 tests in this configuration (using some extra s3tests filter in addition to `@attr('encryption')`)
> 2. outside of `2-kms/`, always deploy a vault instance and configure `rgw_crypt_sse_s3_backend` accordingly. this way we can always run the sse-s3 tests, along with the sse-kms tests against whatever backend is chosen from `2-kms`. i'm just not sure whether we can configure a single vault instance with transit enabled for sse-s3, but with vault_kv.yaml for sse-kms
> 
> what do you think @mdw-at-linuxbox?

@cbodley we can run SSE-KMS (transit) and SSE-S3 (this uses transit) tests at the same time with the same vault by using different namespaces.  But, if you wish to test SSE-S3 and SSE-KMS with KV engine that's not possible with the same vault as far as I know.

**Review by priyagpssehgal — COMMENTED** [review]
I have missed out the code, where PUT header has the value AES256 (i.e., to encrypt using AWS:SSE-S3) a single object, then we generate appropriate KEK and encrypt that object. Currently, the code only considers the bucket level configuration for encryption and not object level for SSE-S3. I will add that piece of code in rgw_crypt.cc and commit it.

**Comment by mattbenjamin** [comment]
> 
> 
> I have missed out the code, where PUT header has the value AES256 (i.e., to encrypt using AWS:SSE-S3) a single object, then we generate appropriate KEK and encrypt that object. Currently, the code only considers the bucket level configuration for encryption and not object level for SSE-S3. I will add that piece of code in rgw_crypt.cc and commit it.

Hi @priyagpssehgal I think @mdw-at-linuxbox may have commits already doing that, @mdw-at-linuxbox could you please confirm?

**Comment by priyagpssehgal** [comment]
> > I have missed out the code, where PUT header has the value AES256 (i.e., to encrypt using AWS:SSE-S3) a single object, then we generate appropriate KEK and encrypt that object. Currently, the code only considers the bucket level configuration for encryption and not object level for SSE-S3. I will add that piece of code in rgw_crypt.cc and commit it.
> 
> Hi @priyagpssehgal I think @mdw-at-linuxbox may have commits already doing that, @mdw-at-linuxbox could you please confirm? (I think there should be review comments from @mdw-at-linuxbox here, and also a PR at this point.)

@mattbenjamin I will wait for @mdw-at-linuxbox in that case.

**Comment by cbodley** [comment]
> Currently, the code only considers the bucket level configuration for encryption and not object level for SSE-S3

maybe we could work on test cases for that in the meantime?

**Comment by mdw-at-linuxbox** [comment]
Sorry for not responding sooner - got bogged down in some customer cases.

I do have some further work but probably nothing directly relevant to the above.  Mainly I have vault interface code that will be able to create and delete keys with somewhat more flexibility than you've got.  My goal is to be able to manage per-bucket keys -- I know that's not important for your use case, but I think for most sites that will be the expected choice.

The sse-s3 logic should be orthogonal to the kms case.  I'm not sure the code in this PR does it, but in theory it ought to be possible to do kms in barbican + sse-s3 in vault.  Testing-wise, the simple case would be to just treat sse-s3 as another environment, bring up vault separately for it, etc.  The more complicated solution would be to allow the cross-product of different kms choices + sse-s3 cases.  Setting up vault is pretty quick, but I don't know that it's worth the combinatorial explosion to try to hit all thost combinations.  Note that even in the base code, there is a degenerate version of sse-s3 that uses a single fixed "test" key in the ceph configuration.

**Comment by cbodley** [comment]
added a TODO list to the PR description to track the remaining work

**Comment by cbodley** [comment]
i pushed a fresh rebase of this branch to https://github.com/cbodley/ceph/commits/pr-42522

**Comment by cbodley** [comment]
i opened https://github.com/ceph/ceph/pull/44339 that adds the teuthology changes necessary to run the sse-s3 test cases from https://github.com/ceph/s3-tests/pull/420

**Comment by cbodley** [comment]
@prasakri @priyagpssehgal Marcus opened https://github.com/ceph/ceph/pull/44494 that builds on this PR, we'd love your help with review

**Comment by priyagpssehgal** [comment]
> @prasakri @priyagpssehgal Marcus opened #44494 that builds on this PR, we'd love your help with review

Sure @cbodley, I will review the PR.

**Pull request closed without merging**